specs/merge/beacon-chain.md

```diff
@@ -69,15 +67,59 @@ This patch adds transaction execution to the beacon chain as part of the Merge f
 #### `BeaconBlockBody`
 
 ```python
-class BeaconBlockBody(phase0.BeaconBlockBody):
+class BeaconBlockBody(Container):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data  # Eth1 data vote
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
     # Execution
     execution_payload: ExecutionPayload  # [New in Merge]
 ```
 
 #### `BeaconState`
 
 ```python
-class BeaconState(phase0.BeaconState):
+class BeaconState(Container):
```

Comment:
ditto, why `class BeaconState(altair.BeaconState)` doesn't work?

Reply:
Without this change, the linter cannot figure out the type of `state.slot` well enough:

```
. venv/bin/activate; cd ./tests/core/pyspec; \
flake8  --config /home/proto/projects/eth2.0-specs/linter.ini ./eth2spec \
&& mypy --config-file /home/proto/projects/eth2.0-specs/linter.ini -p eth2spec.phase0 -p eth2spec.altair -p eth2spec.merge
eth2spec/merge/minimal.py:2238: error: Value of type variable "SupportsLessThanT" of "max" cannot be "object"
eth2spec/merge/mainnet.py:2238: error: Value of type variable "SupportsLessThanT" of "max" cannot be "object"
Found 2 errors in 2 files (checked 9 source files)
make: *** [Makefile:127: lint] Error 1
```

The python line `2238` (in `process_sync_aggregate`):

```
    previous_slot = max(state.slot, Slot(1)) - Slot(1)
```

I think the issues is that the extended type is in a different module, which declares a different `Slot` type. And  mypy cannot handle that with `max`, even though the Slot types both inherit `int`, and interchangeable everywhere else.

Annotating all fields to use types of the merge model works, but at that point you may as well not extend the base-type.

In the long-term, I would prefer to migrate over the eth2 specs to a python-first approach like the new eth1-specs, and generate the documentation from the code. That way we can get rid of code-generation, and have a single type of everything, and using mypy may get a lot easier.

Reply:
> In the long-term, I would prefer to migrate over the eth2 specs to a python-first approach like the new eth1-specs, and generate the documentation from the code.

Agreed!

> That way we can get rid of code-generation, and have a single type of everything, and using mypy may get a lot easier.

By single type, do you mean like `BeaconStateType` or a more general abstract `SszType`? IIUC, to address the typing problem, I guess mypy-izing `remarkable` is required before the overhauling work?

---

Another thing, I found that when using remarkable inheritance, I always worry about if we wrongly inherited the old container where the fields have been updated in the new fork. e.g., *IF* `BeaconBlockHeader` fields are changed. Writing down all the fields may be more reliable in this context. 😅 👍

Reply:
> By single type, do you mean like ...

Single definition. E.g. we repeat common types like `Slot`, `DepositData` etc. in the `phase0`, `altair`, and `merge` packages. 

> I guess mypy-izing remarkable is required before the overhauling work?

I think it's more of a type conflict than an annotation problem with mypy. Suggestions to improve typing in remerkleable are welcome, but that can go into a separate PR.

> I always worry about if we wrongly inherited the old container ...

It should leave fields in the same order, but override the type. Any additional fields are appended at the end. But agreed, repeating the fields, and not forcing class inheritance structures into the spec, is better.